corehq/apps/users/audit/change_messages.py

```diff
@@ -1,91 +1,294 @@
         else:
-            message = "Primary location: None"
-        return message
-
-    @staticmethod
-    def commcare_user_assigned_locations_info(location_names):
-        return f"Assigned locations: {location_names}"
+            change_message = {"location": [
+                {"slug": "clear_primary_location"}
+            ]}
+        return change_message
 
     @staticmethod
-    def web_user_assigned_locations_info(locations_info):
-        return f"Assigned locations: {locations_info}"
+    def assigned_locations_info(locations):
+        if locations:
+            change_message = {"assigned_locations": [
+                {
+                    "slug": "set_assigned_locations",
+                    "params": {"locations": [
+                        {'id': location.location_id, 'name': location.name}
+                        for location in locations
+                    ]}
+                }
+            ]}
+        else:
+            change_message = {"assigned_locations": [
+                {"slug": "clear_assigned_locations"}
+            ]}
+        return change_message
 
     @staticmethod
-    def groups_info(groups_info):
-        return f"Groups: {groups_info}"
+    def groups_info(groups):
+        if groups:
+            change_message = {"groups": [
+                {
+                    "slug": "set_groups",
+                    "params": {"groups": [
+                        {'id': group.get_id, 'name': group.name}
+                        for group in groups
+                    ]}
+                }
+            ]}
+        else:
+            change_message = {"groups": [
+                {"slug": "clear_groups"}
+            ]}
+        return change_message
 
     @staticmethod
     def added_as_web_user(domain):
-        return f"Added as web user to domain '{domain}'"
+        return {"domain": [
+            {
+                "slug": "add_as_web_user",
+                "params": {"name": domain}
+            }
+        ]}
 
     @staticmethod
     def invited_to_domain(domain):
-        return f"Invited to domain '{domain}'"
+        return {"domain_invitation": [
+            {
+                "slug": "add_domain_invitation",
+                "params": {"domain": domain}
+            }
+        ]}
 
     @staticmethod
     def invitation_revoked_for_domain(domain):
-        return f"Invitation revoked for domain '{domain}'"
+        return {"domain_invitation": [
+            {
+                "slug": "remove_domain_invitation",
+                "params": {"domain": domain}
+            }
+        ]}
+
+
+class UserChangeMessageFormatterV1(object):
+    """
+    This class exposes just get_messages method which builds human readable
+    messages for structured messages created by UserChangeMessageV1.
+    """
+    def get_messages(self, change_messages):
+        formatted_messages = []
+        for key, change_message in change_messages.items():
+            if hasattr(self, f"_{key}_messages"):
```

Comment:
this fails the 'grep' test, why not just have the functions use the same name as the slug?

Reply:
Ya. Is there any other way to avoid a very long if else chain to check on slug and call the corresponding method?

One way I thought this was better, by making the methods private, so it was clear they were not called outside and were only accessible via the only public method.

Reply:
another way to do this would be like this:
```python
formatters = {
   "set_program": UserChangeMessageFormatter.simple_formatter(noop("Program: {name}[{id}]")),
   "set_assigned_locations": UserChangeMessageFormatter._set_assigned_locations,   
   "clear_assigned_locations": UserChangeMessageFormatter.simple_formatter(noop("Assigned locations: []")),
}

class UserChangeMessageFormatter:
    @staticmethod
    def simple_formatter(message_format):
        def _formatter(change_details):
            return message_format.format(**change_details.get('params', {}))
        return _formatter

    @staticmethod
    def _set_assigned_locations(change_message):
        return .....

def get_messages(change_messages):
    for change_detail in itertools.chain.from_iterable(change_messages.values()):
        yield formatters[change["slug"]](change_detail)